nodeJS/express/express_forms_validation.md

```diff
@@ -0,0 +1,520 @@
+    <button type="submit" onclick="return confirm('Are you sure you want to delete this user?');">Delete</button>
+  </form>
+  <% }); %>
+</ul>
+```
+
+Then we add the logic to handle the request into our controller:
+
+```javascript
+// Tell the server to delete a matching user, if any. Otherwise, respond with an error.
+exports.users_delete_post = asyncHandler(async (req, res) => {
+  const userId = req.params.userId;
+  if (!userStorage[userId]) return res.status(404).send("User not found.");
+  delete userStorage[userId];
+  res.redirect("/users");
+});
+```
+
+Don't forget to add the new route to your router!
+
+```javascript
+usersRouter.post("/:userId/delete", usersController.users_delete_post);
+```
+
+You'll notice we didn't need a `GET` route here since we're already redirecting to `/users`.
+
+We could go much deeper into working safely with forms, but we'll stop there. By now you can already see how helpful `express-validator` is, and how you can do almost anything you want with the `req.body` object.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+#### Add user details
+
+Expand the `User` model we created to include more details. Implement the following fields and validations to your model:
+
+- Email (required, must be formatted properly)
+- Age (optional, must be a number between 18 and 120)
+- Bio (optional, maximum 200 characters)
+
+Don't forget to update the view to display these new fields!
+
+#### Implement searching
+
+What if we want to search for a specific user in a list of thousands? We'll need a new route and view that lets clients search our list of users.
+
+1. Add a form (in `users.ejs` or another view) which accepts a `name` or `email` (or both!)
+1. Create a new route `/users/search` which accepts `GET` and `POST` requests.
+1. Add the search logic to your controller which searches your list for a matching user.
+   - Your `POST` request should handle searching for the user.
+   - Your `GET` request should then render the search result.
+1. Display the search results in a new view: `search.ejs`.
+
+</div>
+
+### Knowledge check
+
+The following questions are an opportunity to reflect on key topics in this lesson. If you can't answer a question, click on it to review the material, but keep in mind you are not expected to memorize or master this knowledge.
+
+- [What are the attributes of a form element and what do they do?](#html-forms-overview)
+- [How do you install and import express-validator in your project?](#validation-and-sanitization)
+- [How do you validate and sanitize form input using express-validator?](#understanding-the-body-function)
+- [What is the difference between validation and sanitization?](#validation-and-sanitization)
+- [How do you handle validation errors in Express routes?](#validation-results)
+- [What is the importance of escaping HTML characters in a form?](https://owasp.org/www-community/attacks/SQL_Injection)
+
+### Additional resources
```

Comment:
https://express-validator.github.io/docs/guides/customizing#implementing-a-custom-validator

^ custom validators can be an additional resource